### ISSUE TITLE
Introspection

### ISSUE THREAD

**Issue by tiero** [ElementsProject/elements-miniscript#18]
I find very useful for developers to have a way to make use of new fragments that maps to the new introspection opcodes, very useful for making covenants.

DO you have an idea if this fits the purpose of Miniscript and, if yes, what could be a timeline, especially for an initial agreement on the spec? We are implementing in other languages descriptors/miniscript for Elements, so would be cool to work in parallel as soon there could be consensus

**Comment by apoelstra** [comment]
>DO you have an idea if this fits the purpose of Miniscript

Yes -- well, Elements Miniscript at least. Naturally we can't support this in the "real" Miniscript until Bitcoin has some form of introspection ability.

> if yes, what could be a timeline,

My hope was "already" but our upstream re-architecture for Taproot took much longer than expected. Let's say "2 months".

> especially for an initial agreement on the spec?

That will take longer. Initially we'll just stick some new fragments in to solve practical problems that Blockstream and other Liquid developers are trying to solve. It will take some iteration before we're confident that we've gotten the right set of fragments and that we've implemented them optimally.

> We are implementing in other languages descriptors/miniscript for Elements, so would be cool to work in parallel as soon there could be consensus

I think your best bet is to keep up with the current state of this library, provide feedback, yell at us if we change things in ways that break your applications, etc. If you actually deploy something in production using a certain set of fragments we will be careful not to break things after that.

**Comment by tiero** [comment]
My current need is giving a way to users [to load complex script types](https://github.com/vulpemventures/marina/issues/250#issuecomment-1062077859) into Marina Wallet, so it can track balances and spend them if solvable/can sign so to speak.

There is no way to express with current fragments things like CSFS, CAT, new arithmetic opcodes or introspection ones. I understand it will come, so will keep an eye.

For the time being in our JavaScript parser, we are introducing a custom `asm` fragment that let user to load arbitrary opcodes so we can mix them with current fragments. 

I know that this is useless since makes impossible to analyze etc.. but we only want a way to share these opcodes in a formal way, and use descriptor/miniscript when possible alrrady.

> If you actually deploy something in production using a certain set of fragments we will be careful not to break things after that.

Will try to write here the [synthetic asset smart contract](https://github.com/vulpemventures/synthetic-asset-smart-contract) and propose some new fragment for new opocodes so maybe we can make some early thoughts.

**Comment by apoelstra** [comment]
>For the time being in our descriptor parser, we are introducing a custom asm fragment that let user to load arbitrary opcodes so we can mix them with current fragments.

>I know that this is useless since makes impossible to analyze etc.. but we only want a way to share these opcodes in a formal way, and use descriptor/miniscript when possible already.

Be careful with this! It's not only the `asm` fragment itself that is "impossible to analyze" but the entire Miniscript that contains it. But this is likely a reasonable way to proceed for research & development.

With Simplicity we will be able to provide `asm` in the main library since Simplicity's type system can be leveraged to prevent arbitrary code inside one fragment from messing up code in other fragments, without requiring we have a fixed curated set of fragments.

**Comment by tiero** [comment]
> "impossible to analyze" but the entire Miniscript that contains it

Yep! Totally understood.  User will know this is a very reckless feature and they need to be 100% sure of the asm code they are loading.

The idea is just to have a way to share opcodes at runtime for script path spends and implicitly tell the type of address shall be generated. We only parse the descriptor, detect the top level fragment and if a CHECKSIG op is there and a pubkey template, a way for wallet to fill it with his own.

As soon all the new opcodes will be mapped to Miniscript, will likely drop this `asm`fragment and maybe at that point use this library directly  (maybe compiled to WASM to run in the browser if you will provide bindings)

**Comment by sanket1729** [comment]
Have you looked at the `Extensions` in the library? If you want to create any new fragments. You can do it downstream by implementing the [Extension](https://github.com/ElementsProject/elements-miniscript/blob/29316707aea464602da53cdd3afb3d1ebe670a16/src/extensions/mod.rs#L31) trait. 

However, in this case, you are responsible for making sure that the fragment can be parsed unambiguously by the miniscript parser.

**Comment by tiero** [comment]
That looks interesting. 
Not that much experience in Rust tho, would wait for the reference implementations and examples eheh :)

We pretty close to finish our TypeScript descriptor parser btw, we added `asm` fragment as temporary solution and we made sure is valid only inside `TREE` of a `eltr` descriptor, new opcodes must be used in tapscripts anyway. Will drop it as soon new fragments are there and/or with extension trait.

**Comment by sanket1729** [comment]
Hello @tiero, introspection support with arithmetic operations is now supported. See https://github.com/ElementsProject/elements-miniscript/blob/master/doc/extension_spec.md for details. Still awaiting #28 to be merged

**Comment by tiero** [comment]
Amazing @sanket1729 

We plan to add support to our [ionio-lang.org](https://ionio-lang.org) for current `elements-miniscript` fragments, to allow to parse miniscript and compile to opcodes, generating an [Ionio Artifact JSON](https://ionio-lang.org/docs/artifact) in the process, so SDK(s) and libraries can easily spend taproot output scripts.
